File: apps/syn-cli/src/syn_cli/commands/repo.py
"""Repository management commands — register, list, show, assign, and observability."""

from __future__ import annotations

from typing import Annotated

import typer
from rich.table import Table

from syn_cli._output import (
    console,
    format_cost,
    format_status,
    format_timestamp,
    format_tokens,
)
from syn_cli.commands._api_helpers import api_get, api_post, build_params

app = typer.Typer(
    name="repo",
    help="Manage repositories and their observability",
    no_args_is_help=True,
)


@app.command("register")
def register(
    org: Annotated[str, typer.Option("--org", "-o", help="Organization ID")],
    url: Annotated[str, typer.Option("--url", "-u", help="Full repo name (owner/repo)")],
    system: Annotated[
        str | None, typer.Option("--system", "-s", help="Assign to system immediately")
    ] = None,
    provider: Annotated[str, typer.Option(help="Provider (github, gitlab, …)")] = "github",
    branch: Annotated[str, typer.Option(help="Default branch")] = "main",
    private: Annotated[bool, typer.Option("--private/--public", help="Private repo")] = False,
    created_by: Annotated[str, typer.Option(help="Creator identifier")] = "cli",
) -> None:
    """Register a repository with an organization."""
    data = api_post(
        "/repos",
        json={
            "organization_id": org,
            "full_name": url,
            "provider": provider,
            "default_branch": branch,
            "is_private": private,
            "created_by": created_by,
        },
    )

    repo_id = data.get("repo_id", "")
    console.print(f"[green]Repo registered:[/green] {repo_id}")
    console.print(f"  Full name: {url}")

    if system:
        try:
            api_post(f"/repos/{repo_id}/assign", json={"system_id": system})
            console.print(f"  [green]Assigned to system:[/green] {system}")
        except (typer.Exit, Exception):
            console.print("  [yellow]Could not assign to system.[/yellow]")


@app.command("list")
def list_repos(
    org: Annotated[
        str | None, typer.Option("--org", "-o", help="Filter by organization ID")
    ] = None,
    system: Annotated[
        str | None, typer.Option("--system", "-s", help="Filter by system ID")
    ] = None,
    unassigned: Annotated[
        bool, typer.Option("--unassigned", help="Only repos without a system")
    ] = False,
) -> None:
    """List registered repositories."""
    params = build_params(organization_id=org, system_id=system)
    if unassigned:
        params["unassigned"] = True
    data = api_get("/repos", params=params)

    repos = data.get("repos", [])
    if not repos:
        console.print("[dim]No repos found.[/dim]")
        return

    table = Table(title=f"Repos ({data.get('total', len(repos))})")
    table.add_column("ID", style="cyan", no_wrap=True)
    table.add_column("Repo")
    table.add_column("Provider")
    table.add_column("Branch")
    table.add_column("System")
    table.add_column("Created")

    for r in repos:
        table.add_row(
            r.get("repo_id", "")[:12],
            r.get("full_name", ""),
            r.get("provider", ""),
            r.get("default_branch", ""),
            r.get("system_id") or "[dim]—[/dim]",
            format_timestamp(str(r.get("created_at") or "")),
        )
    console.print(table)


@app.command("show")
def show(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
) -> None:
    """Show repo details."""
    r = api_get(f"/repos/{repo_id}")

    console.print(f"[bold]Repo:[/bold] {r.get('full_name', '')}")
    console.print(f"  ID:           {r.get('repo_id', '')}")
    console.print(f"  Provider:     {r.get('provider', '')}")
    console.print(f"  Branch:       {r.get('default_branch', '')}")
    console.print(f"  Organization: {r.get('organization_id', '')}")
    if r.get("system_id"):
        console.print(f"  System:       {r['system_id']}")
    console.print(f"  Private:      {'yes' if r.get('is_private') else 'no'}")
    console.print(f"  Created:      {format_timestamp(str(r.get('created_at') or ''))}")


@app.command("assign")
def assign(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
    system: Annotated[str, typer.Option("--system", "-s", help="System ID")],
) -> None:
    """Assign a repo to a system."""
    api_post(f"/repos/{repo_id}/assign", json={"system_id": system})
    console.print(f"[green]Repo assigned:[/green] {repo_id} → system {system}")


@app.command("unassign")
def unassign(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
) -> None:
    """Remove a repo from its system."""
    api_post(f"/repos/{repo_id}/unassign")
    console.print(f"[yellow]Repo unassigned:[/yellow] {repo_id}")


@app.command("health")
def repo_health(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
) -> None:
    """Show health metrics for a repo."""
    d = api_get(f"/repos/{repo_id}/health")

    trend = d.get("trend", "")
    trend_icon = {"improving": "↑", "degrading": "↓", "stable": "→"}.get(str(trend), "")

    console.print(f"[bold]{d.get('repo_full_name', repo_id)} — Health[/bold]")
    console.print(f"  Executions:   {d.get('total_executions', 0)}")
    console.print(f"  Successful:   {d.get('successful_executions', 0)}")
    console.print(f"  Failed:       {d.get('failed_executions', 0)}")
    rate = d.get("success_rate")
    if rate is not None:
        console.print(f"  Success rate: {rate:.0%}")
    console.print(f"  Trend:        {trend_icon} {trend}")
    console.print(f"  Cost (window):{format_cost(d.get('window_cost_usd') or '0')}")
    console.print(f"  Last run:     {format_timestamp(d.get('last_execution_at'))}")


@app.command("cost")
def repo_cost(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
) -> None:
    """Show cost breakdown for a repo."""
    d = api_get(f"/repos/{repo_id}/cost")

    console.print(f"[bold]{d.get('repo_full_name', repo_id)} — Cost[/bold]")
    console.print(f"  Total cost:    {format_cost(d.get('total_cost_usd') or '0')}")
    console.print(f"  Total tokens:  {format_tokens(d.get('total_tokens') or 0)}")
    console.print(f"  Executions:    {d.get('execution_count', 0)}")

    if d.get("cost_by_model"):
        console.print()
        table = Table(title="By Model", show_edge=False)
        table.add_column("Model", style="cyan")
        table.add_column("Cost", justify="right")
        for model, c in sorted(d["cost_by_model"].items(), key=lambda x: x[1], reverse=True):
            table.add_row(model, format_cost(c))
        console.print(table)


@app.command("activity")
def repo_activity(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
    limit: Annotated[int, typer.Option(help="Max entries")] = 50,
    offset: Annotated[int, typer.Option(help="Pagination offset")] = 0,
) -> None:
    """Show recent execution activity for a repo."""
    data = api_get(f"/repos/{repo_id}/activity", params={"limit": limit, "offset": offset})

    entries = data.get("entries", [])
    if not entries:
        console.print("[dim]No activity found.[/dim]")
        return

    table = Table(title=f"Repo Activity ({data.get('total', 0)} total)")
    table.add_column("Execution", style="cyan", no_wrap=True)
    table.add_column("Workflow")
    table.add_column("Status")
    table.add_column("Started")
    table.add_column("Trigger")

    for e in entries:
        table.add_row(
            e.get("execution_id", "")[:12],
            e.get("workflow_name", ""),
            format_status(e.get("status", "")),
            format_timestamp(e.get("started_at")),
            e.get("trigger_source") or "—",
        )
    console.print(table)


@app.command("failures")
def repo_failures(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
    limit: Annotated[int, typer.Option(help="Max failures to show")] = 50,
) -> None:
    """Show recent execution failures for a repo."""
    data = api_get(f"/repos/{repo_id}/failures", params={"limit": limit})

    failures = data.get("failures", [])
    if not failures:
        console.print("[green]No failures found.[/green]")
        return

    table = Table(title=f"Failures ({data.get('total', len(failures))} total)")
    table.add_column("Execution", style="cyan", no_wrap=True)
    table.add_column("Workflow")
    table.add_column("Phase")
    table.add_column("Failed At")
    table.add_column("Error")

    for f in failures:
        table.add_row(
            f.get("execution_id", "")[:12],
            f.get("workflow_name", ""),
            f.get("phase_name") or "—",
            format_timestamp(f.get("failed_at")),
            (f.get("error_message") or "")[:60],
        )
    console.print(table)


@app.command("sessions")
def repo_sessions(
    repo_id: Annotated[str, typer.Argument(help="Repo ID")],
    limit: Annotated[int, typer.Option(help="Max sessions to show")] = 50,
) -> None:
    """Show agent sessions associated with a repo."""
    data = api_get(f"/repos/{repo_id}/sessions", params={"limit": limit})

    sessions = data.get("sessions", [])
    if not sessions:
        console.print("[dim]No sessions found.[/dim]")
        return

    table = Table(title=f"Sessions ({data.get('total', len(sessions))} total)")
    table.add_column("ID", style="cyan", no_wrap=True)
    table.add_column("Execution", no_wrap=True)
    table.add_column("Status")
    table.add_column("Started")
    table.add_column("Tokens", justify="right")
    table.add_column("Cost", justify="right")

    for s in sessions:
        table.add_row(
            s.get("id", "")[:12],
            s.get("execution_id", "")[:12],
            format_status(s.get("status", "")),
            format_timestamp(s.get("started_at")),
            format_tokens(s.get("total_tokens") or 0),
            format_cost(s.get("total_cost_usd") or "0"),
        )
    console.print(table)

File: apps/syn-cli-node/src/commands/repo.ts
/**
 * Repository management commands — CRUD + observability.
 * Port of apps/syn-cli/src/syn_cli/commands/repo.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import { print, printError, printDim, printSuccess } from "../output/console.js";
import { style, BOLD, CYAN, DIM } from "../output/ansi.js";
import { formatCost, formatDuration, formatStatus, formatTimestamp, formatTokens } from "../output/format.js";
import { Table } from "../output/table.js";


function reqRepoId(parsed: ParsedArgs): string {
  const id = parsed.positionals[0];
  if (!id) { printError("Missing repo-id"); throw new CLIError("Missing argument", 1); }
  return id;
}

const registerCommand: CommandDef = {
  name: "register",
  description: "Register a repository",
  options: {
    url: { type: "string", short: "u", description: "Repository URL (owner/repo or full URL)" },
    system: { type: "string", short: "s", description: "Assign to system ID" },
    org: { type: "string", short: "o", description: "Organization ID" },
  },
  handler: async (parsed: ParsedArgs) => {
    const url = parsed.values["url"] as string | undefined;
    if (!url) { printError("Missing --url"); throw new CLIError("Missing option", 1); }

    let org = parsed.values["org"] as string | undefined;
    if (!org) {
      // Auto-select if exactly one organization exists
      const orgsData = unwrap(await api.GET("/organizations"), "List organizations");
      const orgs = orgsData.organizations ?? [];
      if (orgs.length === 1) {
        org = orgs[0]!.organization_id;
        if (!org) {
          printError("Organization ID missing from API response");
          throw new CLIError("Invalid organization data", 1);
        }
        printDim(`Using organization: ${org}`);
      } else if (orgs.length === 0) {
        printError("No organizations found. Create one first with: syn org create");
        throw new CLIError("No organizations", 1);
      } else {
        printError("Multiple organizations found. Specify one with --org");
        throw new CLIError("Missing option", 1);
      }
    }

    const body = {
      full_name: url,
      organization_id: org,
      provider: "github",
      owner: "",
      default_branch: "main",
      provider_repo_id: "",
      installation_id: "",
      is_private: false,
      created_by: "cli",
    };

    const d = unwrap(await api.POST("/repos", { body }), "Register repository");
    printSuccess(`Repository registered: ${d.repo_id}`);
    print(`  Name: ${d.full_name}`);
  },
};

const listCommand: CommandDef = {
  name: "list",
  description: "List repositories",
  options: {
    org: { type: "string", short: "o", description: "Filter by organization" },
    system: { type: "string", short: "s", description: "Filter by system" },
  },
  handler: async (parsed: ParsedArgs) => {
    const d = unwrap(await api.GET("/repos", {
      params: {
        query: {
          organization_id: (parsed.values["org"] as string | undefined) ?? null,
          system_id: (parsed.values["system"] as string | undefined) ?? null,
        },
      },
    }), "List repositories");
    const items = d.repos ?? [];
    if (items.length === 0) { printDim("No repositories found."); return; }

    const table = new Table({ title: "Repositories" });
    table.addColumn("ID", { style: CYAN });
    table.addColumn("Name");
    table.addColumn("System", { style: DIM });

    for (const r of items) {
      table.addRow(
        r.repo_id,
        r.full_name,
        r.system_id || "\u2014",
      );
    }
    table.print();
  },
};

const showCommand: CommandDef = {
  name: "show",
  description: "Show repository details",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const d = unwrap(await api.GET("/repos/{repo_id}", { params: { path: { repo_id: id } } }), "Get repository");
    print(`${style("Repository:", BOLD)} ${d.full_name}`);
    print(`  ID:     ${d.repo_id}`);
    if (d.system_id) print(`  System: ${d.system_id}`);
    if (d.organization_id) print(`  Org:    ${d.organization_id}`);
  },
};

const assignCommand: CommandDef = {
  name: "assign",
  description: "Assign a repository to a system",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  options: {
    system: { type: "string", short: "s", description: "System ID to assign to" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const system = parsed.values["system"] as string | undefined;
    if (!system) { printError("Missing --system"); throw new CLIError("Missing option", 1); }
    unwrap(await api.POST("/repos/{repo_id}/assign", { params: { path: { repo_id: id } }, body: { system_id: system } }), "Assign repository");
    printSuccess(`Repository ${id} assigned to system ${system}.`);
  },
};

const unassignCommand: CommandDef = {
  name: "unassign",
  description: "Remove repository from its system",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    unwrap(await api.POST("/repos/{repo_id}/unassign", { params: { path: { repo_id: id } } }), "Unassign repository");
    printSuccess(`Repository ${id} unassigned from system.`);
  },
};

const healthCommand: CommandDef = {
  name: "health",
  description: "Show health metrics for a repository",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const d = unwrap(await api.GET("/repos/{repo_id}/health", { params: { path: { repo_id: id } } }), "Get repo health");

    print(`${style("Repository Health:", BOLD)} ${id}`);
    print(`  Success rate:    ${(d.success_rate * 100).toFixed(1)}%`);
    print(`  Trend:           ${d.trend}`);
    print(`  Total runs:      ${d.total_executions}`);
    print(`  Window cost:     ${formatCost(d.window_cost_usd)}`);
    print(`  Window tokens:   ${formatTokens(d.window_tokens)}`);
    if (d.last_execution_at) print(`  Last execution:  ${formatTimestamp(d.last_execution_at)}`);
  },
};

const costCommand: CommandDef = {
  name: "cost",
  description: "Show cost breakdown for a repository",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const d = unwrap(await api.GET("/repos/{repo_id}/cost", { params: { path: { repo_id: id } } }), "Get repo cost");

    print(`${style("Repository Costs:", BOLD)} ${id}`);
    print(`  Total cost:  ${formatCost(d.total_cost_usd)}`);
    print(`  Tokens:      ${formatTokens(d.total_tokens)}`);

    const byModel = d.cost_by_model ?? {};
    const modelEntries = Object.entries(byModel);
    if (modelEntries.length > 0) {
      const table = new Table({ title: "Cost by Model" });
      table.addColumn("Model", { style: CYAN });
      table.addColumn("Cost", { align: "right" });
      for (const [model, cost] of modelEntries) {
        table.addRow(model, formatCost(cost));
      }
      table.print();
    }
  },
};

const activityCommand: CommandDef = {
  name: "activity",
  description: "Show recent execution activity for a repository",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  options: {
    limit: { type: "string", short: "n", description: "Max results", default: "20" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const limit = Number((parsed.values["limit"] as string | undefined) ?? "20");
    const d = unwrap(await api.GET("/repos/{repo_id}/activity", { params: { path: { repo_id: id }, query: { limit } } }), "Get repo activity");
    const items = d.entries ?? [];
    if (items.length === 0) { printDim("No recent activity."); return; }

    const table = new Table({ title: `Activity: ${id}` });
    table.addColumn("Execution", { style: DIM });
    table.addColumn("Workflow");
    table.addColumn("Status");
    table.addColumn("Started");
    table.addColumn("Duration", { align: "right" });

    for (const e of items) {
      table.addRow(
        e.execution_id.slice(0, 12),
        e.workflow_name,
        formatStatus(e.status),
        formatTimestamp(e.started_at),
        e.duration_seconds ? formatDuration(e.duration_seconds * 1000) : "\u2014",
      );
    }
    table.print();
  },
};

const failuresCommand: CommandDef = {
  name: "failures",
  description: "Show recent execution failures for a repository",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  options: {
    limit: { type: "string", short: "n", description: "Max results", default: "10" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const limit = Number((parsed.values["limit"] as string | undefined) ?? "10");
    const d = unwrap(await api.GET("/repos/{repo_id}/failures", { params: { path: { repo_id: id }, query: { limit } } }), "Get repo failures");
    const items = d.failures ?? [];
    if (items.length === 0) { printDim("No recent failures."); return; }

    const table = new Table({ title: `Failures: ${id}` });
    table.addColumn("Execution", { style: DIM });
    table.addColumn("Workflow");
    table.addColumn("Error");
    table.addColumn("Time");

    for (const e of items) {
      table.addRow(
        e.execution_id.slice(0, 12),
        e.workflow_name,
        e.error_message.slice(0, 60) || "\u2014",
        formatTimestamp(e.failed_at),
      );
    }
    table.print();
  },
};

const sessionsCommand: CommandDef = {
  name: "sessions",
  description: "Show agent sessions for a repository",
  args: [{ name: "repo-id", description: "Repository ID", required: true }],
  options: {
    limit: { type: "string", short: "n", description: "Max results", default: "20" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqRepoId(parsed);
    const limit = Number((parsed.values["limit"] as string | undefined) ?? "20");
    const d = unwrap(await api.GET("/repos/{repo_id}/sessions", { params: { path: { repo_id: id }, query: { limit } } }), "Get repo sessions");
    const items = d.sessions ?? [];
    if (items.length === 0) { printDim("No sessions found."); return; }

    const table = new Table({ title: `Sessions: ${id}` });
    table.addColumn("Session", { style: DIM });
    table.addColumn("Status");
    table.addColumn("Started");
    table.addColumn("Tokens", { align: "right" });
    table.addColumn("Cost", { align: "right" });

    for (const s of items) {
      table.addRow(
        s.id.slice(0, 12),
        formatStatus(s.status),
        formatTimestamp(s.started_at),
        formatTokens(s.total_tokens),
        formatCost(s.total_cost_usd),
      );
    }
    table.print();
  },
};

export const repoGroup = new CommandGroup("repo", "Manage repositories and view observability data");
repoGroup
  .command(registerCommand)
  .command(listCommand)
  .command(showCommand)
  .command(assignCommand)
  .command(unassignCommand)
  .command(healthCommand)
  .command(costCommand)
  .command(activityCommand)
  .command(failuresCommand)
  .command(sessionsCommand);
